File: 1456-maximum-number-of-vowels-in-a-substring-of-given-length/1456-maximum-number-of-vowels-in-a-substring-of-given-length.rb
# @param {String} s
# @param {Integer} k
# @return {Integer}
def max_vowels(s, k)
    hash={'a'=>true, 'e'=>true, 'i'=>true, 'o'=>true, 'u'=>true}
    count=j=i=0
    max_count=0
    while i<s.length
        count+=1 if hash[s[i]]
        if i>=k
            count-=1 if hash[s[j]]
            j+=1
        end
        max_count=count if count>max_count
        i+=1
    end
    max_count
end

File: 1456-maximum-number-of-vowels-in-a-substring-of-given-length/1456-maximum-number-of-vowels-in-a-substring-of-given-length.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var maxVowels = function(s, k) {
    let count = 0
    const vowels = new Set(['a','e','i','o','u']);
    let i=0;
    while(i<k){
        if(vowels.has(s.charAt(i)))
            count++
        i++
    }

    let maxVowel = count;

    while(i<s.length && maxVowel<k){
        if(vowels.has(s.charAt(i)))
            count++
        if(vowels.has(s.charAt(i-k)))
            count--
        if(count>maxVowel)
           maxVowel = count 
        i++
    }
    return maxVowel
};